### ISSUE TITLE
allow limited mutation

### ISSUE THREAD

**Issue by ooovi** [DiffMu/DiffPrivacyInferenceHs#71]
idea so far:
- there is a builtin container type (e.g. for model/parameters, wrapping Flux `Params`) that can be mutated by builtin functions whose sensitivity/privacy we control (e.g. for subtracting a gradient from the parameters)
- non-builtin functions that mutate some of their parameters must be annotated with `!` and calls to them cannot be assigned to (i.e. `bla = foo!(a,b,c)` is forbidden)
- we have two new dmterms, namely "lambdas that mutate some of their arguments" and "a call to a mutating lambda followed by some stuff"
- also new dmterms for (for now) the `Params` wrapper and it's builtin friends
- a preprocessing typechecker tracks which of the functions parameters are mutated (only function parameters can be mutated).  it annotates all subterms of the dmterm with types:
```
data MutType : 𝒰₀ where
  Mut   : MutType
  Immut : MutType

data Type : 𝒰₀ where
  Simple : Type
  Arr    : List (Type) × Type -> Type
  MutArr : List (MutType)
```
- new dmterms are generated from this: a term 
```
function update!(model, data)
   ...do something non-mutating with data...
   builtin_update!(model, data)
end
```
gets type ```MutArr [Mut, Immut]``` and hence becomes
```
function update!(model, data)
   ...do something non-mutating with data...
   model = builtin_update!(model, data)
   model
end
```
also every call to `update!(model, data)` becomes `model = update!(model,data)`

**Comment by MxmUrw** [comment]
This is implemented in 53150f61769 and follow up commits. As of 943401bdddaa9 all cases are implemented, but not extensively tested yet.

**Comment by MxmUrw** [comment]
This now has withstood some initial testing and works correctly, but as explained in #95 is currently not safe to use.
